### PR TITLE
feat(backend): submit incident

### PR DESCRIPTION
**Pull request by YogiPristiawan**
Close https://github.com/teknologi-umum/semyi/issues/29

- I am using `x-api-key` header for security. Therefore, you must add `API_KEY` to the environment. Let me know, if you have a better solution.

### CODE REVIEW

**Comment by takadev15** [comment]
This is only for submitting incident right?

**Comment by YogiPristiawan** [comment]
yes

**Comment by takadev15** [comment]
looks good to me 👍🏼

File: backend/http.go [hunk 1]
```diff
@@ -275,3 +283,73 @@ func (s *Server) staticSnapshot(w http.ResponseWriter, r *http.Request) {
...
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte(`{"error": "api key is required"}`))
+		return
+	} else {
+		if apiKey != s.apiKey {
+			w.WriteHeader(http.StatusUnauthorized)
+			w.Header().Set("Content-Type", "application/json")
+			w.Write([]byte(`{"error": "api key is invalid"}`))
+			return
+		}
```
Comment (aldy505) [inline backend/http.go:300]: Won't be this easy for sure. But this will do for now.

File: backend/incident.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		return true
+	}
+	return false
+}
+
+type Incident struct {
+	MonitorID   string           `json:"monitor_id"`
+	Title       string           `json:"title"`
+	Description string           `json:"description"`
+	Timestamp   string           `json:"timestamp"`
```
Comment (aldy505) [inline backend/incident.go:47]: Should've `time.Time`

File: backend/incident.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	Severity    IncidentSeverity `json:"severity"`
+	Status      IncidentStatus   `json:"status"`
+	CreatedBy   string           `json:"created_by"`
+}
+
+func (i Incident) Validate() error {
+	_, err := time.Parse(time.RFC3339, i.Timestamp)
+	if err != nil {
+		return err
+	}
```
Comment (aldy505) [inline backend/incident.go:57]: If `i.Timestamp` is a type of `time.Time`, we should only check if this is non-zero. We will allow this if it's on the future (e.g., for scheduled maintenance job)

File: backend/incident.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+	if !i.Severity.IsValid() {
+		return fmt.Errorf("invalid incident severity")
+	}
+
+	if !i.Status.IsValid() {
+		return fmt.Errorf("invalid incident status")
+	}
```
Comment (aldy505) [inline backend/incident.go:65]: Don't you want to use the `ValidationError` struct I've declared somewhere else?

File: backend/main.go [hunk 5]
```diff
@@ -53,6 +53,11 @@ func main() {
 		port = "5000"
 	}
 
+	apiKey, ok := os.LookupEnv("API_KEY")
+	if !ok {
+		log.Fatal().Msg("API_KEY is required")
```
Comment (aldy505) [inline backend/main.go:58]: Shouldn't be Fatal though. Should be a warning that the `API_KEY` is not set, therefore you can't make API calls to endpoints that requires an API key.

**Review by aldy505 — CHANGES_REQUESTED** [review]

File: backend/incident.go (hunk 4 above)
Comment (YogiPristiawan) [reply backend/incident.go]: [2fd6626](https://github.com/teknologi-umum/semyi/pull/39/commits/2fd6626659a94c0883a22e88b62c6cdb07476680)

File: backend/main.go (hunk 5 above)
Comment (YogiPristiawan) [reply backend/main.go]: [5f24f28](https://github.com/teknologi-umum/semyi/pull/39/commits/5f24f28188dce5593279e5cc4a9ed957a05f2d37)

File: backend/incident_writer.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+func (w *IncidentWriter) Write(ctx context.Context, incident Incident) error {
+	conn, err := w.db.Conn(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get database connection: %w", err)
+	}
+
+	incidentStatus := incident.Status
+	if incident.Timestamp.After(time.Now()) {
+		incidentStatus = IncidentStatusScheduled
+	}
```
Comment (aldy505) [inline backend/incident_writer.go:29]: Nice

**Pull request merged**